### PR TITLE
Recursive Descent PML parser

### PR DESCRIPTION
**Pull request by stmobo**
This is a handwritten [recursive descent parser](https://en.wikipedia.org/wiki/Recursive_descent_parser) for the PML syntax. 

This is _only_ a parser; I don't know how the parser is supposed to instantiate anything, so as of right now the parser does nothing with the data it parses.

Modifying the parser to actually store data would be fairly easy; just use `lastConsumedToken` after successful (true-returning) `consume(Pattern)` calls to get the matched token.

### CODE REVIEW

**Comment by brandon-gong** [comment]
One thing to check: I've slightly tweaked the syntax and added in commas. Hopefully that's just an easy minor fix. Or maybe you are already using this revised syntax, in which case, great.
https://gist.github.com/brandon-gong/b1c003c75832b01f93a50d53aa265708

A few things to consider.
1. I like that I'm seeing the ConfigParseException being used, and I think it should be used more.  Can we
   throw an error when a device/input/filter/pipelinestage/callback string contains an illegal character?
   The regex for that is `->|\||:|;|#|,|\.|\(|\)|\s`.  It's much more convenient to catch this earlier on.
2. How easy would it be to keep track of line and character numbers for more useful error messages?
3. Line 47 of RecursiveDescentParser.java: the current PML syntax does not require the first character to
    be a letter. This is a tiny detail. We will probably be using a letter as the first character anyways.
4. The return value of the parser should be `ArrayList< ArrayList<String> >` or something of the sort.
    basically it should look something like this:

(input)
```text
gamepad1.right_joystick_x
    |->| logging,
    :: always |-> deadband(0.01) -> FOC ->| twist;
```
(output)
```text
[
    ["gamepad1", "right_joystick_x", "changed", "logging"]
    ["gamepad1", "right_joystick_x", "always", "deadband(0.01)", "FOC", "twist"]
]
```
Basically, it's a "flattened" version of the config input, where for each `ArrayList<String>`
```text
index 0: device name
index 1: input name
index 2: filter name
index 3...length-1: pipeline stages
index 4: callback name
```
This makes it so its super easy for the config interpreter to just look at a certain index and search in a
hashmap for that index.

**Comment by brandon-gong** [comment]
If I'm being unclear (which I probably am) just ask.  I'm bad at words.  But it would be nice for you to go ahead and implement some of those things instead of me having to go and figure it out myself.

It is a huge improvement over my own quick and dirty parser though; much less nesting and piling everything into a single method, much more flexibility.  Later on, if we have the need to add things to the config files and extend it past just pipelines, this would be much easier to work with instead of having to write a whole new parser.

**Comment by brandon-gong** [comment]
looks good. I'll probably add in some more comments and stuff later.
